### ISSUE TITLE
How we use the mesh and the operations we need

### ISSUE THREAD

**Issue by jeff-cohere** [RDycore/RDycore#8]
@knepley said (in #6):

> Cool. I think we should make a list of the operations we want to use the mesh in. For example, in CEED we really just want iterations over the dofs, and we autoamtically construct these "restrictions" (list of dof numbers in local vectors). We probably also want automated ways of calculating geometric factors. However, I think most of the code doesn't really need to touch the mesh itself.

Here we can discuss how we use the mesh in our solver code, based on relevant discretizations and a desire to minimize the amount of data that we store and pass around.

**Comment by knepley** [comment]
Excellent. I can start with the "residual evaluation", which I think we should call RE for short, since it comes up all the time. For RE, I think we just need the CEED information, which is dof closures for the integration domains (FV usually has cell pairs), and geometric factors (volumes, normals). I cannot think of anything else there, but maybe @jedbrown can.

**Comment by knepley** [comment]
I wanted to explain the reason for making a serial section before distribution. In order to be able to map the global dof order to the natural order, we must have a natural dof order on the order mesh. Thus we need an initial section.

If we decide to use PetscFV for layout, we do not need to create a Section, just the initial PetscFV.

**Comment by jedbrown** [comment]
Are we planning to use slope reconstruction? I assume yes (but maybe not immediately). For ex11 (which was moved into `PetscFV`), we computed pseudoinverses and placed into face-based storage so slope reconstruction was still a face traversal. Once the cell-centered slopes have been computed, there is another face traversal reconstructing left and right face values and solving Riemann problems. I think to the extent we build matrices for implicit solvers, it would be for the `zero` slope reconstruction, which produces a much sparser matrix and ensures h-ellipticity (and computationally, you skip the first pass).

Probably we should start by expressing Riemann solvers as `CeedQFunction`s. It would have two active inputs (left and right states) plus a passive input providing the face normal (either multiplied by quadrature weight, or with quadrature weight stored separately). @knepley Can we do this in rdycore with plans to migrate it to PETSc, or should we build this as a feature in PETSc?
